src/adr/0031-distributed-mail-queue.md

```diff
@@ -0,0 +1,116 @@
+# 31. Distributed Mail Queue
+
+Date: 2020-04-13
+
+## Status
+
+Accepted (lazy consensus)
+
+## Context
+
+MailQueue is a central component of SMTP infrastructure allowing asynchronous mail processing. This enables a short 
+SMTP reply time despite a potentially longer mail processing time. It also works as a buffer during SMTP pike workload
+to not overload a server. 
+
+Furthermore, when used as a Mail Exchange server (MX), the ability to add delays to be observed before dequeing elements
+allows, among others:
+
+ - Delaying retries upon MX delivery failure to a remote site.
+ - Throttling, which could be helpful for not being considered a spammer.
+
+A mailqueue also enable advanced administration operation like traffic review, discarding emails, resetting wait delays, 
+purging the queue, etc.
+
+Spring implementation and non distributed implementations rely on an embedded ActiveMQ to implement the MailQueue. 
+Emails are being stored in a local file system. An administrator wishing to administrate the mailQueue will thus need 
+to interact with all its James servers, which is not friendly in a distributed setup.
+
+Distributed James relies on the following third party softwares (among other):
+
+ - **RabbitMQ** for messaging. Good at holding a queue, however some advanced administrative operations can't be 
+implemented with this component alone. This is the case for `browse`, `getSize` and `arbitrary mail removal`.
+ - **Cassandra** is the metadata database. Due to **tombstone** being used for delete, queue is a well known anti-pattern.
+ - **ObjectStorage** (Swift or S3) holds large byte content.
+
+## Decision
+
+Distributed James should ship a distributed MailQueue composing the following softwares with the following 
+responsibilities:
+
+ - **RabbitMQ** for messaging. A rabbitMQ consumer will trigger dequeue operations.
+ - A time series projection of the queue content will be maintained in Cassandra (see later). Time series avoid the 
+aforementioned tombstone anti-pattern, and no polling is performed on this projection.
+ - **ObjectStorage** (Swift or S3) holds large byte content. This avoids overwhelming over softwares.
+ 
+Here are details of the Cassandra MailQueue View data-model:
+
+ - **enqueuedMailsV3** holds the time series. The primary key holds the queue name, the (rounded) time of enqueue 
+designed as a slice, and a bucket count. The bucketCount enables sharding and avoids all writes at a given point in 
+time to go to the same Cassandra partition. The clustering key is composed of an enqueueId, a unique identifier. The 
+content holds the metadata of the email. This table enables, from a starting date, to load all of the emails that 
+have ever been in the mailQueue. Content is never deleted.
```

Comment:
s/`Content is never deleted`/`its content is never deleted` ? Or more clear: `any record in this table will never be deleted, instead, a new record is inserted into deletedMailsV2 on each delete operation`

Reply:
s/Content is never deleted/its content is never deleted

:+1:

> any record in this table will never be deleted, instead, a new record is inserted into deletedMailsV2 on each delete operation

That's not what I meant; this is explained on the next line.

I meant to explain the limitation that content is never removed.